src/branch/context.rs

```diff
@@ -19,44 +22,55 @@ pub struct Context {
     ///   - there was neither a project or the --instance option,
     ///   - the project has no linked instance.
     current_branch: Option<String>,
+
+    /// Project dir if it was resolved by instance_name or current directory
+    project_dir: Option<PathBuf>,
 }
 
 impl Context {
     pub async fn new(options: &Options) -> anyhow::Result<Context> {
         // use instance name provided with --instance
         let instance_name = options.conn_params.get()?.instance_name();
-        let instance_name = instance_name.map(|n| n.clone().into());
-        let mut instance_name = instance_name.map(assume_local_instance).transpose()?;
+        let mut instance_name: Option<InstanceName> = instance_name.map(|n| n.clone().into());
+        let project_dir = get_project_dir(None, true).await?;
+        let mut branch: Option<String> = None;
 
-        // fallback to instance name of the associated project
         if instance_name.is_none() {
-            let project_dir = get_project_dir(None, true).await?;
-
-            let name = if let Some(project_dir) = project_dir {
+            instance_name = if let Some(project_dir) = project_dir.as_ref() {
                 let stash_dir = project::stash_path(&fs::canonicalize(project_dir)?)?;
                 project::instance_name(&stash_dir).ok()
             } else {
                 None
             };
-
-            if let Some(name) = name {
-                instance_name = Some(assume_local_instance(name)?);
-            }
         }
 
-        // try to read current branch from instance credentials
-        let current_branch = if let Some(instance_name) = &instance_name {
-            let credentials_path = credentials::path(instance_name)?;
-            let credentials = credentials::read(&credentials_path).await?;
+        // read from credentials
+        if project_dir.is_some()
+            && instance_name.as_ref().map_or(false, |v| match v {
+                InstanceName::Local(_) => true,
+                _ => false,
+            })
+        {
+            let instance_name = match instance_name.as_ref().unwrap() {
+                InstanceName::Local(instance) => instance,
+                unsupported => anyhow::bail!(format!("Cannot use instance type {}", unsupported)),
```

Comment:
A better error message is needed:

```suggestion
                unsupported => anyhow::bail!(format!("cannot switch branches on Cloud instance {org}/{slug}: instance is not linked to a project", unsupported)),
```

Reply:
Wait, why does this apply to _all_ branch commands?  Shouldn't only the `switch` be affected?